# Datan analysointi ja visualisointi TTC8040
## DA&VI Harjoitus 4


Nämä kysymykset ovat suoraan Moodlesta, mutta kurssiin kuuluu myös ohjaustilaisuus ja muistio jokaisesta harjoituskerrasta jossa tulee hyviä vinkkejä tehtävien suorittamiseen.

Neljännen viikon [materiaalia](https://student.labranet.jamk.fi/~pelju/datanav_k22/topic4/kerta4.ipynb). (Tehtävät saattavat sisältää asioita muistakin.)

***Tentti sulkeutuu keskiviikko, 19. huhtikuuta 2023, 23:59.***

---


## Kysymys 1
# Autodatan yhdistäminen
Tee funktio `autot(url1, url2)`, joka tekee järjestyksessä seuraavaa:

* lataa datat DataFrameihin url1:sta ja url2:sta, joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen autoihin liittyvät datat url1:stä ja url2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* tekee juuri yhdistetystä DataFramesta Serieksen, josta käy ilmi kuinka monta automerkkiä datassa on (automerkki on sarakkeessa 'Cn'),
* tulostaa em. Serieksestä viisi eniten esiintyvää automerkkiä.

Tässä tehtävässä:

url1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
url2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv

(Viittaus [autotA1](https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv),
sekä [autotA2](https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv);
plus testissä [autotA3](https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv) sekä [autotA4](https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv).)

### Lopuksi pitäisi näyttää joksenkin tältä

    OCTAVIA    396
    FABIA      377
    RAPID      197
    SUPERB     159
    KAROQ      148
    Name: Cn, dtype: int64


In [17]:
def autot(url1, url2):
    import pandas as pd
    
    df = pd.read_csv(url1, sep=';')
    df2 = pd.read_csv(url2, sep=';')
    auto_df = pd.concat([df, df2])
    tulos = auto_df['Cn'].value_counts().head()
    print(tulos)
    
### alla olevat rivit vain omaan testaukseen, ei palautukseen
url_T1_1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_T1_2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'

url_T1_3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv'
url_T1_4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv'

autot(url_T1_1,url_T1_2)
#autot(url_T1_3,url_T1_4)

# PS tehtävässä pitää esiintyä pandas

OCTAVIA    396
FABIA      377
RAPID      197
SUPERB     159
KAROQ      148
Name: Cn, dtype: int64


## Kysymys 2
# Autodatan yhdistäminen ja ryhmittely
Tee funktio `autot(url_a1, url_a2, url_b1, url_b2)`, joka tekee järjestyksessä seuraavaa:

* lataa datat DataFrameihin osoitteista `(url_a1, url_a2, url_b1, url_b2)`, joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos DataFramessa 'df' on sarakkeessa auton moottorin teho 'ep (KW)' nan arvoja, korvataan ne lukuarvolla nolla,
* tuottaa DataFramen 'tulos', johon on laskettu autojen keskimääräiset tehot sarakkeesta 'ep (KW)' ryhmiteltynä auton valmistajan mukaan 'Mh' (tutki tarvittaessa groupby [dokumentaatiosta](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html?highlight=groupby#pandas.DataFrame.groupby) kuinka groupby palauttaa indeksin),
* järjestää DataFramen 'tulos' rivit keskimääräisen tehon mukaan 'ep (KW)' suurimmasta pienimpään.
* tulostaa edellä jäjestetystä DataFramesta viisi ensimmäistä riviä.

Lisäksi Teamsissä oli esitetty kysymys kuinka pitäisi käsitellä tuplariviä. Automaattitarkistaja näytti edellyttävän, ettei käytetä `drop_duplicates` vaan ne jätettäisiin paikoilleen, mutta tähän ei toistaiseksi ollut opettajalta kommenttia.

Tässä tehtävässä:

url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

(Lisäksi testeissä näyttäisi olevan käytössä myös

url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv
url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv
url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv
url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv
)

### Lopuksi pitäisi näyttää joksenkin tältä

                    Mh     ep (KW)
    6            TESLA  386.000000
    0           AA-IVA  267.750000
    2  FORD WERKE GMBH   89.741007
    5            SKODA   81.372515
    3              KIA   77.000000

In [33]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    
    df1 = pd.read_csv(url_a1, sep= ';')
    df2 = pd.read_csv(url_a2, sep= ';')
    df3 = pd.read_csv(url_b1, sep= ';')
    df4 = pd.read_csv(url_b2, sep= ';')
    
    dfa = pd.concat([df1, df2])
    dfb = pd.concat([df3, df4])

    df = pd.merge(dfa, dfb, on = 'ID')
    df['ep (KW)'].fillna(inplace=True, value=0)   
    tulos = df.groupby('Mh', as_index=False)['ep (KW)'].mean()
    tulos = tulos.sort_values('ep (KW)', ascending = False).head()
    print(tulos)
    
### alla olevat rivit vain omaan testaukseen, ei palautukseen
url_T2_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_T2_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_T2_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_T2_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

url_T2_a3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv'
url_T2_a4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv'
url_T2_b3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv'
url_T2_b4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv'

autot(url_T2_a1, url_T2_a2, url_T2_b1, url_T2_b2)
# autot(url_T2_a3, url_T2_a4, url_T2_b3, url_T2_b4)

# PS tehtävässä pitää esiintyä pandas

                Mh     ep (KW)
6            TESLA  386.000000
0           AA-IVA  267.750000
2  FORD WERKE GMBH   89.741007
5            SKODA   81.372515
3              KIA   77.000000


## Kysymys 3
# Autodatan yhdistäminen ja ristiintaulukointi
Tee funktio `autot(url_a1, url_a2, url_b1, url_b2)`, joka tekee järjestyksessä seuraavaa:

* lataa datat DataFrameihin osoitteista (url_a1, url_a2, url_b1, url_b2), joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos sarakkeessa 'Ft' on merkkijono 'DIESEL', korvataan se merkkijonolla 'Diesel',
* jos sarakkeessa 'Ft' on merkkijono 'PETROL', korvataan se merkkijonolla 'Petrol',
* tekee uuden DataFramen hyödyntäen [crosstab funktiota](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) ja sopivaa numpy [funktiota](https://numpy.org/doc/stable/reference/), jossa on laskettu esimerkkituloksen mukaisesti keskimääräiset päästöt sarakkeesta 'Enedc (g/km)' autonvalmistaja kohtaisesti (sarake 'Mh') polttoainetyypeittäin (sarake 'Ft'),
* esittää edellä kuvatusta DataFramesta keskimääräiset päästöt yhden desimaalin tarkkuudella,
* valitsee uuteen DataFrameen sarakkeet 'Dielsel' ja 'Petrol',
* tulostaa edellä tuotetusta DataFramesta viisi ensimmäistä riviä.

Tässä tehtävässä:

url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

(Lisäksi näissäkin testeissä näyttäisi olevan käytössä myös

url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv
url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv
url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv
url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv
)

### Lopuksi pitäisi näyttää joksenkin tältä
	
    Ft                   Diesel  Petrol
    Mh
    AA-IVA                129.0   194.7
    AUTOMOBILES CITROEN   102.7   105.6
    FORD WERKE GMBH       112.9   133.7
    KIA                     NaN    88.0
    NISSAN                112.3   127.7

In [28]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    import numpy as np
    
    df1 = pd.read_csv(url_a1, sep= ';')
    df2 = pd.read_csv(url_a2, sep= ';')
    df3 = pd.read_csv(url_b1, sep= ';')
    df4 = pd.read_csv(url_b2, sep= ';')
    
    dfa = pd.concat([df1, df2])
    dfb = pd.concat([df3, df4])

    df = pd.merge(dfa, dfb, on = 'ID')
    df['Ft'].replace(['DIESEL', 'PETROL'], ['Diesel', 'Petrol'], inplace=True)
    new_df = pd.crosstab(df['Mh'],  df['Ft'], values=df['Enedc (g/km)'], aggfunc=np.mean).round(1)
    tulos= new_df[['Diesel', 'Petrol']].head()

    print(tulos)
    
### alla olevat rivit vain omaan testaukseen, ei palautukseen
url_T3_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_T3_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_T3_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_T3_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

url_T3_a3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv'
url_T3_a4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv'
url_T3_b3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv'
url_T3_b4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv'

autot(url_T3_a1, url_T3_a2, url_T3_b1, url_T3_b2)
#autot(url_T3_a3, url_T3_a4, url_T3_b3, url_T3_b4)

# PS tehtävässä pitää esiintyä pandas

Ft                   Diesel  Petrol
Mh                                 
AA-IVA                129.0   194.7
AUTOMOBILES CITROEN   102.7   105.6
FORD WERKE GMBH       112.9   133.7
KIA                     NaN    88.0
NISSAN                112.3   127.7


## Kysymys 4
# Autodatan yhdistäminen ja ryhmittely monitasoiseksi DataFrameksi

Tee funktio `autot(url_a1, url_a2, url_b1, url_b2)`, joka tekee järjestyksessä seuraavaa:

* lataa datat DataFrameihin osoitteista `(url_a1, url_a2, url_b1, url_b2)`, joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos sarakkeessa 'Ft' on merkkijono 'DIESEL', korvataan se merkkijonolla 'Diesel',
* jos sarakkeessa 'Ft' on merkkijono 'PETROL', korvataan se merkkijonolla 'Petrol',
* tekee uuden DataFramen, jossa on laskettu moottorien keskiarvotilavuudet (sarake 'ec (cm3)') ryhmiteltynä sarakkeiden 'Mh' ja 'Ft' mukaan,
* esittää moottorien keskiarvotilavuudet yhden desimaalin tarkkuudella,
* muodostaa monitasoisen DataFramen, jossa sarakeindekseinä ovat monikot ('ec (cm3)', 'Diesel') ja ('ec (cm3)', 'Petrol'),
* jos DataFramessa esiintyy NaN arvoja, niin DataFramen soluun kirjoitetaan merkkijono "no value",
* tulostaa vastaukseen 10 ensimmäistä riviä.

Tässä tehtävässä:

url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

(Lisäksi näissäkin testeissä näyttäisi olevan käytössä myös

url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv
url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv
url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv
url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv
)

### Lopuksi pitäisi näyttää joksenkin tältä

                         ec (cm3)
    Ft                     Diesel    Petrol
    Mh
    AA-IVA                 1995.0    3596.0
    AUTOMOBILES CITROEN    1560.0    1131.0
    FORD WERKE GMBH        1615.1    1291.5
    KIA                  no value    1580.0
    NISSAN                 1559.4    1247.5
    SKODA                  1799.5    1190.6
    TESLA                no value  no value

In [74]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    import numpy as np
    
    df1 = pd.read_csv(url_a1, sep= ';')
    df2 = pd.read_csv(url_a2, sep= ';')
    df3 = pd.read_csv(url_b1, sep= ';')
    df4 = pd.read_csv(url_b2, sep= ';')
    
    dfa = pd.concat([df1, df2])
    dfb = pd.concat([df3, df4])

    df = pd.merge(dfa, dfb, on = 'ID') 
    df['Ft'].replace(['DIESEL', 'PETROL'], ['Diesel', 'Petrol'], inplace=True)
    
    # Lasketaan keskiarvotilavuudet groupby- ja mean-funktioiden avulla
    df2 = df.groupby(['Mh', 'Ft'])['ec (cm3)'].mean().round(1)
    df3 = df2.unstack()
    df3 = df3.fillna('no value')

    tulos = df3[['Diesel', 'Petrol']].head(10)
    tulos.columns = pd.MultiIndex.from_tuples([('ec (cm3)', 'Diesel'), ('ec (cm3)', 'Petrol')], names=['','Ft'])

    print(tulos)  
    
### alla olevat rivit vain omaan testaukseen, ei palautukseen
url_T4_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_T4_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_T4_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_T4_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

url_T4_a3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv'
url_T4_a4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv'
url_T4_b3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv'
url_T4_b4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv'

autot(url_T4_a1, url_T4_a2, url_T4_b1, url_T4_b2)
#autot(url_T4_a3, url_T4_a4, url_T4_b3, url_T4_b4)

# PS tehtävässä pitää esiintyä pandas

                     ec (cm3)          
Ft                     Diesel    Petrol
Mh                                     
AA-IVA                 1995.0    3596.0
AUTOMOBILES CITROEN    1560.0    1131.0
FORD WERKE GMBH        1615.1    1291.5
KIA                  no value    1580.0
NISSAN                 1559.4    1247.5
SKODA                  1799.5    1190.6
TESLA                no value  no value


## Kysymys 5
# Autodatan yhdistäminen ja ryhmittely monitasoiseksi DataFrameksi (2)

Tee funktio `autot(url_a1, url_a2, url_b1, url_b2)`, joka tekee järjestyksessä seuraavaa:

* lataa datat DataFrameihin osoitteista (url_a1, url_a2, url_b1, url_b2), joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos sarakkeessa 'Ft' on merkkijono 'DIESEL', korvataan se merkkijonolla 'Diesel',
* jos sarakkeessa 'Ft' on merkkijono 'PETROL', korvataan se merkkijonolla 'Petrol',
* tallentaa uuteen DataFrameen sarakkeet 'Enedc (g/km)', 'MS' ja 'Ft',
* korvaa sarakkeen nimen 'Enedc (g/km)' nimellä 'Avg emissions (g/km)',
* korvaa sarakkeen nimen 'MS' nimellä 'Member State',
* korvaa sarakkeen nimen 'Ft' nimellä 'Fuel type',
* laskee uuteen DataFrameen keskimääräiset päästöt (sarake 'Avg emissions (g/km)') maittain (sarake 'Member State') ja polttoainetyypeittäin (sarake 'Fuel type'),
* jättää edellä tuotettuun monitasoiseen DataFrameen vain 'Diesel' ja 'Petrol' sarakkeet,
* laskee uuteen sarakkeeseen 'D-P' kuinka paljon sarkkeen 'Diesel' arvo on suurempi kuin sarakkeen 'Petrol',
* järjestää DataFramen sarakkeen 'D-P' mukaan laskevaan järjestykseen,
* pyöristää ja esittää DataFramen lukuarvot kokonaislukujen tarkkuudella,
* tulostaa vastaukseen DataFramesta viisi ensimmäistä riviä.

Tässä tehtävässä taas ihan samat kahdeksan linkkiä kuin kolmessa yllä olevassa tehtävässä niin saa ihan vapaasti katsoa sieltä.

Lisävinkkinä Teamsissa opettaja neuvoi tällaista koodia lukuarvojen INTiksi pyöristämiseen:

    df.round(decimals=0).applymap("{:.0f}".format)

### Lopuksi pitäisi näyttää joksenkin tältä

                 Avg emissions (g/km)
    Fuel type                  Diesel Petrol D-P
    Member State
    EE                            132    108  25
    FI                            137    112  25
    LV                            131    108  23
    GR                            122    100  22
    CZ                            122    110  12

In [47]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    import numpy as np
    
    df1 = pd.read_csv(url_a1, sep= ';')
    df2 = pd.read_csv(url_a2, sep= ';')
    df3 = pd.read_csv(url_b1, sep= ';')
    df4 = pd.read_csv(url_b2, sep= ';')
    
    dfa = pd.concat([df1, df2])
    dfb = pd.concat([df3, df4])

    df = pd.merge(dfa, dfb, on = 'ID') 
    df['Ft'].replace(['DIESEL', 'PETROL'], ['Diesel', 'Petrol'], inplace=True)
    
    df= df[['Enedc (g/km)', 'MS', 'Ft']]
    df.rename(columns = {'Enedc (g/km)':'Avg emissions (g/km)', 'MS':'Member State', 'Ft': 'Fuel type'}, inplace=True)
    df = df.groupby(['Member State', 'Fuel type'])['Avg emissions (g/km)'].mean()
    df = df.unstack()
    df = df[['Diesel','Petrol']]
    df.columns = pd.MultiIndex.from_tuples([('Avg emissions (g/km)', 'Diesel'), ('Avg emissions (g/km)', 'Petrol')], names=['','Fuel type'])
    df['Avg emissions (g/km)','D-P'] = df['Avg emissions (g/km)', 'Diesel'] - df ['Avg emissions (g/km)','Petrol']
    df = df.sort_values(('Avg emissions (g/km)','D-P'), ascending = False)
    df = df.fillna(0)
    tulos = df.round().astype(int).head()
    
    print(tulos)

    
### alla olevat rivit taas vain omaan testaukseen, ei palautukseen
url_T5_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_T5_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_T5_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_T5_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

url_T5_a3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA3.csv'
url_T5_a4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA4.csv'
url_T5_b3='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB3.csv'
url_T5_b4='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB4.csv'

autot(url_T5_a1, url_T5_a2, url_T5_b1, url_T5_b2)
#autot(url_T5_a3, url_T5_a4, url_T5_b3, url_T5_b4)

# PS tehtävässä pitää yllättäen esiintyä pandas

             Avg emissions (g/km)           
Fuel type                  Diesel Petrol D-P
Member State                                
EE                            132    108  25
FI                            137    112  25
LV                            131    108  23
GR                            122    100  22
CZ                            122    110  12
